models/hymod/include/Hymod.h

```diff
@@ -117,25 +126,57 @@ class hymod_kernel
         double et = calc_et(soil_m, et_params);
 
         // get the slow flow output for this time - ks
-        double slow_flow_out = ks_fluxes.slow_flow_in;
+        double slow_flow = groundwater.response(slow);
 
         for(unsigned long int i = 0; i < nash_cascade.size(); ++i)
         {
             runoff = nash_cascade[i].response(runoff);
         }
 
         // record all fluxs
-        fluxes.slow_flow_in = slow;
-        fluxes.slow_flow_out = slow_flow_out;
+        fluxes.slow_flow = slow_flow;
         fluxes.runnoff = runoff;
         fluxes.et_loss = et;
 
         // update new state
         new_state.storage = soil_m - et;
+        new_state.groundwater_storage = groundwater.get_storage();
         for ( unsigned long i = 0; i < nash_cascade.size(); ++i )
         {
             new_state.Sr[i] = nash_cascade[i].get_storage();
         }
+
+        return mass_check(params, state, input_flux, new_state, fluxes);
+
+    }
+
+    static int mass_check(const hymod_params& params, const hymod_state& current_state, double input_flux, const hymod_state& next_state, const hymod_fluxes& calculated_fluxes)
+    {
+        // initalize both mass values from current and next states storage
+        double inital_mass = current_state.storage + current_state.groundwater_storage;
+        double final_mass = next_state.storage + next_state.groundwater_storage;
+
+        // add the masses of the reservoirs before and after the time step
+        for ( int i = 0; i < params.n; ++i)
+        {
+            inital_mass += current_state.Sr[i];
+            final_mass += next_state.Sr[i];
+        }
+
+        // increase the inital mass by input value
+        inital_mass += input_flux;
+
+        // increase final mass by calculated fluxes
+        final_mass += (calculated_fluxes.et_loss + calculated_fluxes.runnoff + calculated_fluxes.slow_flow);
+
+        if ( inital_mass - final_mass > 0.000001 )
```

Comment:
This tolerance should ultimately be a little more dynamic, perhaps a macro to define application wide mass balance tolerance?

Reply:
That was a first cut

Reply:
according to my get this did not add new files just modify existing files.

Reply:
New commit removed the duplicate files